### ISSUE TITLE
Firewall Shared Service

### ISSUE THREAD

**Issue by Lybecker** [microsoft/AzureTRE#882]
**Description**
As a TRE service integrator
I want an uncomplicated way of enable egress traffic to a Workspace/Workspace Service
So that I can create a bundle that can access the required resources

As an Azure administrator
I want to know which firewall rules belongs to which Workspace/Workspace Service/Shared Service 
So that I can get an overview of which rules are needed

As an Azure administrator
I want to be able to create rules manually without interfering with how TRE works
So that I can add custom rules to the firewall

_A future requirement_
As an Azure administrator
I want to be able to approve any new egress rules
So that I can manage which rules are added to the firewall

**Solution design**

Workspace templates must only change resources it itself creates. Changes done to management/core services such as Azure Firewall, Shared Service such as source mirror, Azure Active Directory etc. must be delegated.

Built another microservice, that is responsible for updating the Azure Firewall, which one of the shared services in the TRE.

The Azure Firewall rules comes in two flavors, and we need to support both:

- Application rules: Configure fully qualified domain names (FQDNs) that can be accessed from a subnet.
- Network rules: Configure rules that contain source addresses, protocols, destination ports, and destination addresses.

When a Workspace/Workspace Service needs egress traffic, it must output the requirements as part of the Porter output.

This the flow:

1. The Workspace/Workspace Service Porter bundles output the requirement of a new egress rule like below:

```JSON
{
    ...
    "appEgressRules": [
        // A specific resources firewall app rules.
        {
            "sourceIp": "<10.0.0.1>|<192.168.0.1-192.168.0.5>|<10.2.7.0/24>", //must only be allowed to specify own address space
            "protocol": "TCP"
            "destinationPorts": 
            [
              443,
              80
            ]
            "targetFqdns": [
                "bing.com",
                "msn.com"
            ]
        }
    ],
    "networkEgressRules": [
        // A specific resources firewall app rules.
        {
            "sourceIp": "<10.0.0.1>|<192.168.0.1-192.168.0.5>|<10.2.7.0/24>", //must only be allowed to specify own address space
            "protocol": "TCP"
            "destinationPorts": 
            [
              443,
              80
            ]
            "serviceTags": [
                "AzureContainerRegistry"
            ]
            "targetFqdns": [
                "bing.com",
                "msn.com"
            ]
        }
    ],
},
```

2. The Resource Processor forwards these rules to the Firewall Shared Service via a message. 
3. The Firewall Shared Service validates the rules including if the rules only changes egress rules for the specific Workspace (source IP).
4. The Firewall Shared Service updates the Azure Firewall with the new rules. It created a rule collection per Workspace (the priority is based on existing rule collections + 1).
5. After the Firewall is updated the Firewall Shared Service sends a message to the API about the rules are updated.

For this to work, the Firewall needs to keep its own state (rule ownership) and request information from the API (IP address space).
The Workspace state in the State Store must also have another property store when the Firewall rules have been applied.

_Consideration_: Having the Firewall Shared Service also manage Application Gateway rules for inbound routed traffic.

**Comment by marrobi** [comment]
@Lybecker I thought we had this covered in #23 - along with implementation steps - just hadn't been progressed yet? Let's discuss tomorrow.

**Comment by marrobi** [comment]
Closing as taken merged into #23
